### ISSUE TITLE
Not able to upload any files using @theming-controlpanel-mapper

### ISSUE THREAD

**Issue by NicolasGoeddel** [plone/plone.app.theming#165]
There is an issue with uploading files when `plone.rest` is installed. I always get a 404 error when trying to upload a file. But I am able to create an empty file an copy paste the content to it.

The guys from [plone.rest](https://github.com/plone/plone.rest/issues/59) told me it's your fault because [themefile.py](https://github.com/plone/plone.app.theming/blob/master/src/plone/app/theming/browser/themefile.py) is using "application/json" responses every time instead of the proper status codes 200 and 422 (unprocessable entity).

Please can you fix that issue so we are able to upload files again?

At the moment we are using plone.app.theming == 4.0.1, plone.rest == 1.4.0 and plone.restapi == 4.3.1.

**Comment by djay** [comment]
This is incorrect. The problem is caused as plone.rest won't allow any other part of plone to use apis that expect JSON as a response and expect all other parts of plone to not expect JSON responses.

**Comment by NicolasGoeddel** [comment]
I only want to make it work again. I don't know what's the best way to do it. I don't care who did what wrong. ;-) I hope you guys will find a solution.

**Comment by djay** [comment]
There is already a working fix. https://github.com/plone/plone.rest/pull/61. We use it in production.

**Comment by ewohnlich** [comment]
I've not been able to implement this. I have the following monkey patch and I verified with a debug trace that it was hitting the expected condition and return obj, but it doesn't actually put anything in portal_resources

```def restPublishTraverse(self, request, name):
    try:
        obj = super(RESTTraverse, self).publishTraverse(request, name)
        if not IContentish.providedBy(obj) and not IService.providedBy(obj):
            if isinstance(obj, VirtualHostMonster):
                return obj
            elif isinstance(obj, BTreeFolder2) and obj.id == 'portal_resources':
                return obj
            else:
                raise KeyError
    except KeyError:
        return self._
old_restPublishTraverse(request, name)```

I get a 200 response back in the browser

**Comment by tisto** [comment]
Just for the record. This method in plone.app.theming is causing the problem:

https://github.com/plone/plone.app.theming/blob/master/src/plone/app/theming/browser/themefile.py#L25

The __call__ method clearly violates HTTP/REST standards and best practices because:

- it always returns the HTTP status code 200 (it should return 3xx if an error occurred -> HTTP standard violation)
- it always returns a body (if you send 200 you should not send a body (otherwise use 204) -> HTTP standard violation)
- it encapsulates the error/success message in the HTTP body (this is what HTTP status codes are for -> HTTP standard violation)

If this function would be implemented properly without violating the core of the HTTP standard, no JSON would need to be sent over the wire and no HTTP accept header "application/json" would need to be sent.

**Comment by petri** [comment]
The front-end code that sets the "application/json" accept header for the upload request is not in this package afaik. Which package has the frontend?

**Comment by ewohnlich** [comment]
The [mockup](https://github.com/plone/mockup) package calls the BrowserView mentioned here ("themeFileUpload") in the filemanager pattern, if that's what you mean.

**Comment by tisto** [comment]
If you do not want to refactor the code too much you can just use a different Accept header, e.g. rename "application/json" to "application/json+whatever". This is a two lines change that will fix the issue. I'd recommend fixing this properly of course but that would be a quick fix.

**Comment by petri** [comment]
Oh well. I could not find where mockup code sets the accept: header. I'm starting to think maybe it's set by default by some frontend library patterns use (e.g. dropzone.js) ...

**Comment by tisto** [comment]
I never used mockup or patternslib, so I have no clue how this works.

**Comment by petri** [comment]
Taking a bit closer look at the patterns/mockup theme editor frontend, "violation of HTTP/REST standards and best practices", as @tisto put it, seems pretty widespread. Many of the theme editor TTW operations seem to use returned JSON success/error values, anyway. 

So improving purity of just upload feature here & in frontend might leave the frontend in inconsistent state - things'd work differently per operation (e.g. upload vs. rename vs. whatever). Unless of course all those other operations were changed as well, including what seems quite a few tests.

If someone can locate where the pattern or whatever underlying library is setting the accept: header as was suggested here, perhaps changing that can make theming play nice with `plone.rest` with less effort.

**Comment by petri** [comment]
@tisto for your information, to change the Accept: header of the request, it seems we'd have to fork **Dropzone.js** . Or monkey-patch it's [94-line _uploadData method](https://gitlab.com/meno/dropzone/-/blob/master/src/dropzone.js#L2014-2108) at run-time.

**Comment by petri** [comment]
see https://github.com/plone/plone.rest/pull/105 that would fix this issue.
